sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/worker/StateSampler.java

```diff
@@ -41,6 +44,92 @@
 @ThreadSafe
 public class StateSampler implements AutoCloseable {
 
+  /**
+   * An interface for scheduling state sampling.
+   */
+  public static interface SampleScheduler {
+    Closeable scheduleSampling(StateSampler sampler, long samplingPeriodMs);
+  }
+
+  /**
+   * An implementation of {@link SampleScheduler} using a scheduled executor service and two
+   * callbacks per sample. The sampling uses a "stratified sampling" is used -- in every bucket of
+   * samplingPeriodMs we choose a random point to sample. This prevents pathological behavior in
+   * case some states happen to occur at a similar period
+   * <p>
+   * The current implementation uses a fixed-rate timer with a period samplingPeriodMs as a
+   * trampoline to a one-shot random timer which fires with a random delay within
+   * samplingPeriodMs.
+   * */
+  private static final class DefaultScheduler implements SampleScheduler {
+    private static final int NUM_EXECUTOR_THREADS = 16;
+    private static final ScheduledExecutorService EXECUTOR_SERVICE =
+        Executors.newScheduledThreadPool(NUM_EXECUTOR_THREADS,
+            new ThreadFactoryBuilder().setDaemon(true).build());
+
+    @Override
+    public Closeable scheduleSampling(final StateSampler sampler, final long samplingPeriodMs) {
+      return new Closeable() {
+        private ScheduledFuture<?> invocationTriggerFuture = null;
+        private ScheduledFuture<?> invocationFuture = null;
+        {
+          invocationTriggerFuture =
+              EXECUTOR_SERVICE.scheduleAtFixedRate(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      long delay = ThreadLocalRandom.current().nextInt((int) samplingPeriodMs);
+                      synchronized (this) {
+                        if (invocationFuture != null) {
+                          invocationFuture.cancel(false);
```

Comment:
I don't think this one is a problem since within the same synchronized block we immediately re-set invocationFuture after cancelling.